pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package infracluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	mapiv1beta1 "github.com/openshift/api/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	cerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ptr "k8s.io/utils/ptr"
+	azurev1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	errUnableToGetAzureMAPIProviderSpec = errors.New("unable to get Azure MAPI ProviderSpec")
+)
+
+const (
+	clusterSecretName               = "capz-manager-cluster-credential"    // #nosec G101
+	capzManagerBootstrapCredentials = "capz-manager-bootstrap-credentials" // #nosec G101
+)
+
+// ensureAzureCluster ensures the AzureCluster cluster object exists.
+//
+//nolint:funlen,cyclop
```

Comment:
Maybe adding these is a sign that you could refactor this function to be a little simpler?

Perhaps some of the stages should be factored out to make the flow easier to understand?

```
ensureClusterSecret()
ensureClusterIdentity()
ensureAzureInfraCluster()
```

Reply:
I've refactored it further, unfortunately it just moved the `//nolint:funlen` down to `ensureAzureInfraCluster()`, but `ensureAzureCluster` should be much tidier now?

Reply:
huh, seems I broke something moving that out to it's own function so will take a look tomorrow